### PR TITLE
rsx: Reimplement projected texturing to play nice with the subimage views

### PR DESCRIPTION
**Pull request by kd-11**
https://github.com/RPCS3/rpcs3/issues/14142

Testing needed:
- [x] Resident Evil: Code Veronica X
- [x] TLOU
- [x] Uncharted 2 & 3

### CODE REVIEW

**Comment by Linear524** [comment]
Hi KD-11 !
I tried this PR and unfortunately RPCS3 crashes when trying to load main menu :(
Logs shows this:
E RSX: ERROR: 0:633: 'coord' : undeclared identifier 
ERROR: 0:633: 'xyz' : vector swizzle selection out of range 
ERROR: 0:633: 'w' : vector swizzle selection out of range 
ERROR: 0:633: '[]' : scalar integer expression required 
ERROR: 0:633: '_texcoord_xform' : no matching overloaded function found 
ERROR: 0:633: 'texture' : no matching overloaded function found 
ERROR: 0:633: '' : compilation terminated 
ERROR: 7 compilation errors. No code generated.

E RSX: 
F {RSX Worker 2} SIG: Thread terminated due to fatal error: Failed to compile fragment shader
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\VKProgramPipeline.cpp:29[:25], in function compile)
E RSX: ERROR: 0:632: 'coord' : undeclared identifier 
ERROR: 0:632: 'xyz' : vector swizzle selection out of range 
ERROR: 0:632: 'w' : vector swizzle selection out of range 
ERROR: 0:632: '[]' : scalar integer expression required 
ERROR: 0:632: '_texcoord_xform' : no matching overloaded function found 
ERROR: 0:632: 'texture' : no matching overloaded function found 
ERROR: 0:632: '' : compilation terminated 
ERROR: 7 compilation errors. No code generated.

E RSX: 
F {RSX Worker 10} SIG: Thread terminated due to fatal error: Failed to compile fragment shader
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\VKProgramPipeline.cpp:29[:25], in function compile)
E RSX: ERROR: 0:634: 'coord' : undeclared identifier 
ERROR: 0:634: 'xyz' : vector swizzle selection out of range 
ERROR: 0:634: 'w' : vector swizzle selection out of range 
ERROR: 0:634: '[]' : scalar integer expression required 
ERROR: 0:634: '_texcoord_xform' : no matching overloaded function found 
ERROR: 0:634: 'texture' : no matching overloaded function found 
ERROR: 0:634: '' : compilation terminated 
ERROR: 7 compilation errors. No code generated.

E RSX: 
F {RSX Worker 1} SIG: Thread terminated due to fatal error: Failed to compile fragment shader
(in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\VKProgramPipeline.cpp:29[:25], in function compile)

File: rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXFragmentTextureOps.glsl [hunk 1]
```diff
@@ -27,36 +27,39 @@ R"(
...
+	#define SHADOW_COORD_PROJ(index, coord4) _texcoord_xform_shadow(coord4.xyz / coord4.w, texture_parameters[index])
 
 	#define TEX2D_SHADOW(index, coord3) texture(TEX_NAME(index), SHADOW_COORD(index, coord3))
 	#define TEX3D_SHADOW(index, coord4) texture(TEX_NAME(index), SHADOW_COORD4(index, coord4))
 	#define TEX2D_SHADOWPROJ(index, coord4) textureProj(TEX_NAME(index), SHADOW_COORD_PROJ(index, coord4))
 #else
 	#define TEX2D_SHADOW(index, coord3) texture(TEX_NAME(index), vec3(COORD_SCALE2(index, coord3.xy), coord3.z))
 	#define TEX3D_SHADOW(index, coord4) texture(TEX_NAME(index), vec4(COORD_SCALE3(index, coord4.xyz), coord4.w))
-	#define TEX2D_SHADOWPROJ(index, coord4) textureProj(TEX_NAME(index), vec4(COORD_SCALE2(index, coord4.xy), coord4.zw))
+	#define TEX2D_SHADOWPROJ(index, coord4) texture(TEX_NAME(index), COORD_PROJ3(index, coord))
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXFragmentTextureOps.glsl:62]:
```suggestion
	#define TEX2D_SHADOWPROJ(index, coord4) texture(TEX_NAME(index), COORD_PROJ3(index, coord4))

```

File: rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXFragmentTextureOps.glsl [hunk 2]
```diff
@@ -27,36 +27,39 @@ R"(
...
+#define TEX1D_PROJ(index, coord4) _process_texel(texture(TEX_NAME(index), COORD_PROJ1(index, coord4.xw)), TEX_FLAGS(index))
 #endif
 
 #ifdef _ENABLE_TEX2D
 #define TEX2D(index, coord2) _process_texel(texture(TEX_NAME(index), COORD_SCALE2(index, coord2)), TEX_FLAGS(index))
 #define TEX2D_BIAS(index, coord2, bias) _process_texel(texture(TEX_NAME(index), COORD_SCALE2(index, coord2), bias), TEX_FLAGS(index))
 #define TEX2D_LOD(index, coord2, lod) _process_texel(textureLod(TEX_NAME(index), COORD_SCALE2(index, coord2), lod), TEX_FLAGS(index))
 #define TEX2D_GRAD(index, coord2, dpdx, dpdy) _process_texel(textureGrad(TEX_NAME(index), COORD_SCALE2(index, coord2), dpdx, dpdy), TEX_FLAGS(index))
-#define TEX2D_PROJ(index, coord4) _process_texel(textureProj(TEX_NAME(index), vec4(COORD_SCALE2(index, coord4.xy), coord4.z, coord4.w)), TEX_FLAGS(index))
+#define TEX2D_PROJ(index, coord4) _process_texel(texture(TEX_NAME(index), COORD_PROJ2(index, coord4)), TEX_FLAGS(index))
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXFragmentTextureOps.glsl:47]: COORD_PROJ2 takes a coord3, shouldn't it be COORD_PROJ3 ?

Comment (kd-11) [reply rpcs3/Emu/RSX/Program/GLSLSnippets/RSXProg/RSXFragmentTextureOps.glsl:47]: Projected variants take one extra input which is the scale of the input vector.

**Comment by kd-11** [comment]
@Linear524 Retest

**Comment by solarmystic** [comment]
Still getting the same fatal error and rpcs3 crash after booting up TLOU using the latest PR build unfortunately.

`F {RSX [0x008590c]} SIG: Thread terminated due to fatal error: Failed to compile fragment shader (in file C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\rpcs3\Emu\RSX\VK\VKProgramPipeline.cpp:29[:25], in function compile) fatal error with your latest pr build`

Logfile:-
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/11990714/RPCS3.log)

**Comment by davewuk** [comment]
There are quite a few permutations to test in the Uncharted games and they have quite significant graphics issues without the use of patches in both this PR and master.

I tested a few combinations as follows:-

1. No upscaling, no SPU post-processing removal patch

- flickering rectangles on title screen

NPEA00364_screenshot_2023_07_08_16_41_48

(the glitched area is at the top in the middle, it's hard to see so compare against this normal screenshot)
NPEA00364_screenshot_2023_07_08_16_41_47

- flickering bloom in-game

This is all as expect as it happens in the current master build too. The game isn't really playable without the SPU post-processing removal patch despite what it says in the patch description.

2. Upscaled to 1080p, no SPU post-processing removal patch

- Title screen is green

NPEA00364_screenshot_2023_07_08_16_50_31

(I didn't test in game but can if you want)

3. Upscaled to 1080p, with SPU post-processing removal patch

- Title screen is normal and in-game bloom doesn't flicker

NPEA00364_screenshot_2023_07_08_16_52_48

On my first playthrough I used the settings from the wiki and set the Resolution Scale Threshold to 160x160 as recommended.I remember that this looked mostly OK but on some levels the bloom was pixellated. It seems to be OK in the current master, I'd have to go back to an earlier build to find out exactly when that got fixed

NPEA00364_screenshot_2023_07_08_17_05_37


Is there anything else you need testing before I move on to Uncharted 3?

**Comment by kd-11** [comment]
@davewuk Thanks. The PR breaks some things from reports on discord, I'll update when the found issues are fixed.

**Comment by davewuk** [comment]
Sure thing - if there is anything in particular you need me to test in UC2 & 3 just let me know

**Comment by readywer** [comment]
GT5 with this pr.
Képernyőfelvétel (360)
Képernyőfelvétel (361)

**Comment by solarmystic** [comment]
TLOU dynamic shadow fixed as of the latest pr commit on my end. 

image

**Comment by davewuk** [comment]
The updated PR appears to fix torches in Uncharted 2 that must have been broken a few builds back as I don't remember it being like that in my playthrough.

1. Current Master

NPEA00364_screenshot_2023_07_09_10_04_02

2. Updated PR

NPEA00364_screenshot_2023_07_09_10_02_18

**Comment by davewuk** [comment]
Actually after testing a little more there are places where the torches still glitch out

NPEA00364_screenshot_2023_07_09_10_33_38

NPEA00364_screenshot_2023_07_09_10_34_15

**Comment by davewuk** [comment]
Moving on to Uncharted 3 - there's a similar improvement in dynamic shadows as Uncharted 2 but the same issue with the patches/upscaling. The MLAA patch has to be on (even if not upscaling) to remove glitches despite what it says in the description. There's also something crazy happening in the scenes with hundreds of spiders that I don't remember from my first playthrough.

1. Master - dynamic shadows are glitched

BCES01175_screenshot_2023_07_09_11_10_44

2. PR - glitches seem to be fixed

BCES01175_screenshot_2023_07_09_11_15_31

3. Flickering dark polygons during spider scenes happen in both Master and this PR

BCES01175_screenshot_2023_07_09_10_47_51

I don't know if these are dynamic shadows or something else?

**Comment by kd-11** [comment]
Try with an older build from like a month ago and see if the problem is fixed there with the same settings.

**Comment by davewuk** [comment]
> Try with an older build from like a month ago and see if the problem is fixed there with the same settings.

It isn't possible to test UC2 with _exactly_ the same settings in earlier builds as I have been using upscaling. To get upscaling to work in, say, #13975 I have to set the threshold to 160x160 or I get the infamous white blobs. With 160x160 the white blobs vanish but the bloom is pixelated.

1. UC2 in build 13975 upscaled to 1080p with threshold at 16x16 (default)

Missing bloom
NPEA00364_screenshot_2023_07_09_14_27_53

White blobs
NPEA00364_screenshot_2023_07_09_14_29_32

Torches (flashlights) look OK though

2.  UC2 in build 13975 upscaled to 1080p with threshold at 160x160 (as recommended in the wiki)

Pixelated bloom
NPEA00364_screenshot_2023_07_09_14_31_21

No white blobs and torches (flashlights) look mostly OK with only minor glitches
NPEA00364_screenshot_2023_07_09_14_32_10

In this PR it is possible to upscale to 1080p with 16x16 and you get better looking bloom, almost correct dynamic lighting and no white blobs.

This is my patch configuration for all tests, otherwise I'm using the recommended settings from the Wiki except where mentioned above.
patch config

Specs:
i7 12700k
16gb DDR4 3200 ram
RTX 2060 super

**Comment by davewuk** [comment]
BCES01175_screenshot_2023_07_09_14_57_58

In uncharted 3 I'm getting a lot of stability issues when testing with upscaling and the game keeps freezing in all the builds I've tried.  However when testing in build 13975 from 9th June I seem to get similar results to the current PR, at least when it works. This is using the settings recommended in the wiki plus the following patches:

patch config

I have also had a issues with the dynamic lighting sometimes not working when logging is disabled. I've logged that as a seperate issue.

**Comment by kd-11** [comment]
@davewuk Is there anything that worked better (with regards to shadows) on the older build? Verify without using any upscaling. If produces correct results, we can proceed with the merge.

**Comment by davewuk** [comment]
> Is there anything that worked better (with regards to shadows) on the older build?

Not really no, but there are improvements in other areas.

In build 15159 the shadows and dynamic lights from the flashlight in UC2 on that level look the same as the build from this PR. However, _even when not upscaling_ build 15159 has broken bloom and white blobs (with the same settings and patches). So for me the PR is a great improvement as it means the game can be played at native resolution without the white blob issue and can be upscaled without setting threshold to 160x160 which breaks the bloom in many chapters.

Build 15159 - missing bloom
NPEA00364_screenshot_2023_07_09_15_58_52

Build 15159 - white blobs and occasionally glitched dynamic shadows
NPEA00364_screenshot_2023_07_09_15_59_57

This PR - correct bloom
NPEA00364_screenshot_2023_07_09_16_10_49

This PR - exactly the same shadow glitches at the same position angle as the older build
NPEA00364_screenshot_2023_07_09_16_11_33

I am having a lot of trouble with stability in UC3 today and its making it frustrating to test but broadly speaking the shadows and dynamic lighting looks the same in this PR as it does in the older build.

> Also check if it can be reproduced stock. Patches may trigger buggy codepaths in the gamecode as they often re-enable things that the developers did not intend to run.

I can try, but the level I'm testing on specifically does not work well withou the GPU lighting patch

**Comment by kd-11** [comment]
Thanks for confirming. I'll proceed and merge this work then.

**Comment by davewuk** [comment]
> Thanks for confirming. I'll proceed and merge this work then.

No worries - FWIW I did try just that part of that level in UC2 with all patches disabled (except one that disables SPU post-processing) and I didn't notice any difference in the PR build.

**Comment by Linear524** [comment]
Uncharted series (2 and 3) was always partially broken. As well as TLoU post processing effects and SPU MLAA.
But anyway - I suggest to never use scaling and use FSR upscaling instead... With native 720p. This is the only way to see all intended graphical effects without those "skipping patches"... Just try to set WCB+RCB+WDB+RDB without MLAA skipping patch and watch the result...

**Comment by Linear524** [comment]
Without WCB+RCB+WDB+RDB and without working MLAA you will miss:
- Color Balance and Dynamic Exposure effects
- DoF effects
- Bloom control
- Morphological Antialiasing
- Proper explosion FXs
- Vignette effects
- Film Grain effect
- Proper reflections on cubemaps
- etc

**Comment by davewuk** [comment]
Can you run tests and prove that?

**Comment by Linear524** [comment]
@davewuk
 - I don't have to, because I saw this for years and you can try it by yourself

**Comment by kd-11** [comment]
Don't add discussion comments to closed PRs.

**Pull request merged**